library/std/src/thread/local.rs

```diff
@@ -260,6 +261,29 @@ macro_rules! __thread_local_inner {
             #[inline]
             fn __init() -> $t { $init }
 
+            // When reading this function you might ask "why is this inlined
+            // everywhere other than MSVC?", and that's a very reasonable
+            // question to ask. The short story is that it segfaults rustc if
+            // this function is inlined. The longer story is that MSVC looks to
+            // not support `extern` references to thread locals across DLL
+            // boundaries. This appears to at least not be supported in the ABI
+            // that LLVM implements.
+            //
+            // Because of this we never inline on MVSC, but we do inline on
+            // other platforms (where external references to thread locals
+            // across DLLs are supported). A better fix for this would be to
+            // inline this function on MSVC, but only for "statically linked"
+            // components. For example if two separately compiled rlibs end up
+            // getting linked into a DLL then it's fine to inline this function
+            // across that boundary. It's only not fine to inline this function
+            // across a DLL boundary. Unfortunately rustc doesn't currently have
+            // this sort of logic available in an attribute, and it's not clear
+            // that rustc is even equipped to answer this (it's more of a Cargo
+            // question kinda). This means that, unfortunately, MSVC gets the
+            // pessimistic path for now where it's never inlined.
+            //
+            // The issue of "should enable on MSVC sometimes" is #84933
+            #[cfg_attr(not(target_env = "msvc"), inline)]
```

Comment:
Where you say "cargo", I would say "linker", or rather "if you are doing link-time-codegen, then it is knowable/doable" but maybe "cargo" is correct.